### ISSUE TITLE
why can we use Bernoulli distribution in decoder?

### ISSUE THREAD

**Issue by egrcc** [y0ast/VAE-TensorFlow#3]
hello, thanks for sharing this code. I am confused about a question. I find MNIST data are continuous(pixel value not just 0 and 1), so why can we use Bernoulli distribution in decoder?

**Comment by staturecrane** [comment]
Binary-cross entropy measures the distance between an input and target, including as probabilities. The sigmoid activation squashes the image values between 0 -1, which, as far as I know, means they can then be interpreted as probabilities. Hence, cross-entropy can be used between the output and the target values.

**Comment by egrcc** [comment]
@staturecrane thanks for your reply. I know binary cross-entropy can measure the distance. I just confused about the loss implemented in this code(line 65~70) which I think is not the same with the loss in the paper(https://arxiv.org/pdf/1312.6114v10.pdf).

In this code, the loss is
```python
KLD = -0.5 * tf.reduce_sum(1 + logvar_encoder - tf.pow(mu_encoder, 2) - tf.exp(logvar_encoder), reduction_indices=1)

x_hat = tf.matmul(hidden_decoder, W_decoder_hidden_reconstruction) + b_decoder_hidden_reconstruction
BCE = tf.reduce_sum(tf.nn.sigmoid_cross_entropy_with_logits(x_hat, x), reduction_indices=1)

loss = tf.reduce_mean(BCE + KLD)
```
and in the paper, the loss is 

![default](https://cloud.githubusercontent.com/assets/6469953/20590038/8bd6ae12-b25b-11e6-9cd1-b196343cb83c.PNG)

The KLD is the same with the first term of right, and when p(x|z) is Bernoulli distribution, it is true that the BCE is the same with the second term of right. When p(x|z) is Gaussian distribution, the BCE is not the same. However, only when x is a discrete variable, can we use Bernoulli distribution to model p(x|z). So my question is, what exactly the distribution of p(x|z) is in this code when x is a continuous variable, or do I misunderstand the code or the paper?

the Bernoulli MLP in the paper is

![2](https://cloud.githubusercontent.com/assets/6469953/20590141/5a70a714-b25c-11e6-8dc6-4ac1b3dc94a6.PNG)

**Comment by y0ast** [comment]
So most MNIST pixels are actually 0 or 1, so it is approximately binary. There are fully binarized versions of mnist (by Hugo Larochelle) which indeed given a better score when trained on.

In general this is so close to being binary that it should be ok. It definitely works better than when trained using the continuous loss function.

FYI in the paper also BCE is used for the MNIST experiments

**Comment by jprobichaud** [comment]
I know I'm late to the party, but this paper just got publish and provide some important conclusions: 
The continuous Bernoulli: fixing a pervasive error in variational autoencoders
https://arxiv.org/pdf/1907.06845.pdf

**Comment by SHINJIYOUN** [comment]
sigmoid activation function makes these values between 0 to 1. So the results of multivariate bernoulli distribution assumption is came with blurred gray mnist image.
